Unreliable test timings when using data.draw with text strategy

Using `strategies.data` to draw `strategies.text` consistently raises `hypothesis.errors.Flaky` due to `Unreliable test timings` for minimal examples such as `''`

Minimal example:
```python
@given(st.data())
def test_hypothesis_flaky_text(data):
    data.draw(st.text())
    assert 1
```

Produces:
```
============================================== FAILURES ==============================================
_____________________________________ test_hypothesis_flaky_text _____________________________________

    @given(st.data())
>   def test_hypothesis_flaky_text(data):

natural_merge_sort.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/ben/.local/lib/python3.7/site-packages/hypothesis/core.py:592: in execute
    % (test.__name__, text_repr[0])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hypothesis.core.StateForActualGivenExecution object at 0x7fbc41de3c88>
message = 'Hypothesis test_hypothesis_flaky_text(data=data(...)) produces unreliable results: Falsified on the first call but did not on a subsequent one'

    def __flaky(self, message):
        if len(self.falsifying_examples) <= 1:
>           raise Flaky(message)
E           hypothesis.errors.Flaky: Hypothesis test_hypothesis_flaky_text(data=data(...)) produces unreliable results: Falsified on the first call but did not on a subsequent one

/home/ben/.local/lib/python3.7/site-packages/hypothesis/core.py:751: Flaky
--------------------------------------------- Hypothesis ---------------------------------------------
Falsifying example: test_hypothesis_flaky_text(data=data(...))
Draw 1: ''
Unreliable test timings! On an initial run, this test took 902.74ms, which exceeded the deadline of 200.00ms, but on a subsequent run it took 0.04 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.
==================================== 1 failed, 1 skipped in 1.00s ====================================
```
Using `strategies.text` directly with `@given` **does not** produce the error. The following passes consistently:
```
@given(st.text())
def test_hypothesis_flaky_text(txt):
    assert 1
```

## System Information
- Python 3.7.3
- Hypothesis 4.36.2
- pytest 5.1.2
- Ubuntu 19.04

Happy to look into a fix if this can be determined to be a real problem and not an isolated issue :+1:

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,8 @@
+RELEASE_TYPE: patch
+
+This patch fixes :issue:`2108`, where the first test using
+:func:`~hypothesis.strategies.data` to draw from :func:`~hypothesis.strategies.characters`
+or :func:`~hypothesis.strategies.text` would be flaky due to unreliable test timings.
+
+Time taken by lazy instantiation of strategies is now counted towards drawing from
+the strategy, rather than towards the deadline for the test function.

--- a/hypothesis-python/src/hypothesis/internal/conjecture/data.py
+++ b/hypothesis-python/src/hypothesis/internal/conjecture/data.py
@@ -830,16 +830,20 @@ class ConjectureData(object):
                 % (strategy,)
             )
 
+        at_top_level = self.depth == 0
+        if at_top_level:
+            # We start this timer early, because accessing attributes on a LazyStrategy
+            # can be almost arbitrarily slow.  In cases like characters() and text()
+            # where we cache something expensive, this led to Flaky deadline errors!
+            # See https://github.com/HypothesisWorks/hypothesis/issues/2108
+            start_time = benchmark_time()
+
         if strategy.is_empty:
             self.mark_invalid()
 
         if self.depth >= MAX_DEPTH:
             self.mark_invalid()
 
-        return self.__draw(strategy, label=label)
-
-    def __draw(self, strategy, label):
-        at_top_level = self.depth == 0
         if label is None:
             label = strategy.label
         self.start_example(label=label)
@@ -849,7 +853,6 @@ class ConjectureData(object):
             else:
                 try:
                     strategy.validate()
-                    start_time = benchmark_time()
                     try:
                         return strategy.do_draw(self)
                     finally:

--- a/hypothesis-python/tests/cover/test_health_checks.py
+++ b/hypothesis-python/tests/cover/test_health_checks.py
@@ -27,6 +27,7 @@ from hypothesis import HealthCheck, given, settings
 from hypothesis.control import assume
 from hypothesis.errors import FailedHealthCheck, InvalidArgument
 from hypothesis.internal.compat import int_from_bytes
+from hypothesis.searchstrategy.lazy import LazyStrategy
 from hypothesis.searchstrategy.strategies import SearchStrategy
 from tests.common.utils import checks_deprecated_behaviour, no_shrink
 
@@ -198,3 +199,24 @@ def test_hung_test_is_deprecated():
 def test_it_is_an_error_to_suppress_non_healthchecks():
     with raises(InvalidArgument):
         settings(suppress_health_check=[1])
+
+
+slow_down_init = True
+
+
+def slow_init_integers(*args, **kwargs):
+    # This mimics st.characters() or st.text(), which perform some
+    # expensive Unicode calculations when the cache is empty.
+    global slow_down_init
+    if slow_down_init:
+        time.sleep(0.5)  # We monkeypatch time, so this is fast
+        slow_down_init = False
+    return st.integers(*args, **kwargs)
+
+
+@given(st.data())
+def test_lazy_slow_initialization_issue_2108_regression(data):
+    # Slow init in strategies wrapped in a LazyStrategy, inside an interactive draw,
+    # should be attributed to drawing from the strategy (not the test function).
+    # Specificially, this used to fail with a DeadlineExceeded error.
+    data.draw(LazyStrategy(slow_init_integers, (), {}))

